bentoml/_internal/bento/build_config.py

```diff
@@ -54,98 +59,211 @@ def _convert_python_version(py_version: t.Optional[str]) -> t.Optional[str]:
-                    self.base_image,
-                    self.distro,
+                    f"docker base_image {self.base_image} is used, "
+                    f"'distro={self.distro}' option is ignored.",
                 )
             if self.python_version is not None:
                 logger.warning(
-                    "docker base_image %s is used, 'python=%s' option is ignored",
-                    self.base_image,
-                    self.python_version,
+                    f"docker base_image {self.base_image} is used, "
+                    f"'python={self.python_version}' option is ignored.",
                 )
-            if self.gpu is not None:
+            if self.cuda_version is not None:
                 logger.warning(
-                    "docker base_image %s is used, 'gpu=%s' option is ignored",
-                    self.base_image,
-                    self.gpu,
+                    f"docker base_image {self.base_image} is used, "
+                    f"'cuda_version={self.cuda_version}' option is ignored.",
                 )
 
-    def with_defaults(self) -> "DockerOptions":
-        # Convert from user provided options to actual build options with default values
-        update_defaults = {}
+    def _validate_options(self) -> None:
+        if self.cuda_version is not None:
+            if self.distro not in CUDA_SUPPORTED_DISTRO:
+                raise BentoMLException(
+                    f"{self.distro} does not support CUDA. "
+                    f"Supported distros that have CUDA supports are: {', '.join(CUDA_SUPPORTED_DISTRO)}."
+                )
 
-        if self.base_image is None:
-            if self.distro is None:
-                update_defaults["distro"] = DOCKER_DEFAULT_DISTRO
-            if self.python_version is None:
-                update_defaults["python_version"] = PYTHON_MINOR_VERSION
-            if self.gpu is None:
-                update_defaults["gpu"] = False
+        distro_info = DistroSpec.from_distro(
+            self.distro, cuda=self.cuda_version not in (None, "")
+        )
+
+        if self.python_version is not None:
+            if self.python_version not in distro_info.python_version:
+                raise BentoMLException(
+                    f"{self.python_version} is not supported for {self.distro}. "
+                    f"Supported python versions are: {', '.join(distro_info.python_version)}."
+                )
 
-        return attr.evolve(self, **update_defaults)
+        if self.cuda_version is not None:
+            if self.cuda_version != "default" and (
+                self.cuda_version not in distro_info.cuda_version
+            ):
+                raise BentoMLException(
+                    f"{self.cuda_version} is not supported for "
+                    f"{self.distro}. Supported cuda versions are: {', '.join(distro_info.cuda_version)}."
+                )
+
+    @property
+    def default_values(self) -> t.Dict[str, t.Any]:
```

Comment:
I think the original `with_defaults` is a cleaner design and more flexible, since this doesn't precisely return default_values, but the updates to `self` to fill in missing options.  it can be easily misunderstood for getting the entire default values map.